dev/com.ibm.ws.artifact.zip/src/com/ibm/ws/artifact/zip/cache/internal/ZipFileReaper.java

```diff
@@ -149,6 +149,52 @@ public void validate() {
 
     //
 
+    // Locking note:
+    //
+    // 'Tr.warning' can cause a bundle load, which can cause a call to 'AppClassLoader.loadClass'.
+    //
+    // But, 'AppClassLoader.loadClass' can obtain an entry's input stream, which causes
+    // a call to 'ZipFileReaper.open'.
+    //
+    // See below for stacks which show typical deadlocked threads.
+    //
+    // To avoid the deadlock, the stall warning is issued from a new thread.
+    //
+    // zip file reaper 55 BLOCKED
+    // com.ibm.ws.classloading.internal.AppClassLoader.loadClass(AppClassLoader.java:480)
+    // java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:1032)
+    // java.base/java.util.ResourceBundle$Control.newBundle(ResourceBundle.java:3170)
+    // java.base/java.util.ResourceBundle.loadBundle(ResourceBundle.java:1994)
+    // ...
+    // java.base/java.util.ResourceBundle.getBundle(ResourceBundle.java:1509)
+    // ...
+    // com.ibm.ws.logging.internal.impl.BaseTraceService.warning(BaseTraceService.java:512)
+    // com.ibm.websphere.ras.Tr.warning(Tr.java:653)
+    // com.ibm.ws.artifact.zip.cache.internal.ZipFileReaper$ReaperRunnable.run(ZipFileReaper.java:235)
+    // 
+    // Default Executor-thread-2 40 BLOCKED
+    // com.ibm.ws.artifact.zip.cache.internal.ZipFileReaper.open(ZipFileReaper.java:1018)
+    // ...
+    // com.ibm.ws.artifact.zip.internal.ZipFileEntry.getInputStream(ZipFileEntry.java:176)
+    // com.ibm.ws.classloading.internal.ContainerClassLoader$ArtifactEntryUniversalResource.getByteResourceInformation(ContainerClassLoader.java:517)
+    // ...
+    // com.ibm.ws.classloading.internal.AppClassLoader.loadClass(AppClassLoader.java:482)
+    // java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:1032)
+    // org.apache.derby.impl.sql.compile.SQLParser.insertColumnsAndSource(SQLParser.java:8249)
+    // ...
+    // com.ibm.ws.rsadapter.jdbc.WSJdbcStatement.executeUpdate(WSJdbcStatement.java:608)
+    // ...
+    // javax.servlet.http.HttpServlet.service(HttpServlet.java:575)
+
+    protected static void asyncWarning(final String msgKey, final Object ... msgArgs) {
+        Runnable warningAction = new Runnable() {
+            public void run() {
+                Tr.warning(tc, msgKey, msgArgs);
+            }
+        };
+        (new Thread(warningAction)).start();
```

Comment:
Should there be a dedicated thread like we have for the reaper thread?

Reply:
Modified to launch a new daemon thread which handles async warnings.  See the new commit.